### PR TITLE
[tests] Allow tests to pass when stderr is non-empty

### PR DESCRIPTION
**Pull request by jnewbery**
Resurrect #10241 with nits addressed

Not sure how much people want this. Would be useful for functional tests which cause bitcoind to print to stderr.

### CODE REVIEW

File: test/functional/test_runner.py [hunk 1]
```diff
@@ -40,17 +39,28 @@
     CROSS = "x "
     CIRCLE = "o "
 
+# Default colors to empty strings.
+BOLD, BLUE, RED, GREY, MAGENTA = [("", "")] * 5
```
Comment (jimmysong) [inline test/functional/test_runner.py]: nit: alpha sort other than BOLD?

File: test/functional/test_runner.py [hunk 2]
```diff
@@ -40,17 +39,28 @@
...
 if os.name == 'posix':
     # primitive formatting on supported
     # terminal via ANSI escape sequences:
     BOLD = ('\033[0m', '\033[1m')
-    BLUE = ('\033[0m', '\033[0;34m')
-    RED = ('\033[0m', '\033[0;31m')
     GREY = ('\033[0m', '\033[1;30m')
+    RED = ('\033[0m', '\033[0;31m')
+    BLUE = ('\033[0m', '\033[0;34m')
+    MAGENTA = ('\033[0m', '\033[0;35m')
```
Comment (jimmysong) [inline test/functional/test_runner.py]: nit: alpha sort?

File: test/functional/test_runner.py [hunk 3]
```diff
@@ -40,17 +39,28 @@
...
 
 TEST_EXIT_PASSED = 0
 TEST_EXIT_SKIPPED = 77
 
+STATUS_PASSED = "Passed"
+STATUS_PASSED_WITH_WARNINGS = "Passed with warnings"
+STATUS_SKIPPED = "Skipped"
+STATUS_FAILED = "Failed"
+STATUSES = [STATUS_PASSED, STATUS_PASSED_WITH_WARNINGS, STATUS_SKIPPED, STATUS_FAILED]
```
Comment (jimmysong) [inline test/functional/test_runner.py]: This doesn't seem to be used anywhere else, maybe combine with line below?

**Review by jimmysong — COMMENTED** [review]
utACK, a few nits.

File: test/functional/test_runner.py (hunk 3 above)
Comment (MeshCollider) [inline test/functional/test_runner.py]: IMO its cleaner how it is, shouldn't be combined

File: test/functional/test_runner.py [hunk 4]
```diff
@@ -422,17 +436,20 @@ def __init__(self, name, status, time):
...
+        if self.status == STATUS_PASSED:
             color = BLUE
             glyph = TICK
-        elif self.status == "Failed":
-            color = RED
-            glyph = CROSS
-        elif self.status == "Skipped":
+        if self.status == STATUS_PASSED_WITH_WARNINGS:
+            color = MAGENTA
+            glyph = TICK
```
Comment (MeshCollider) [inline test/functional/test_runner.py]: Is TICK the best glyph to use for passed with warnings? Or something more neutral?

**Comment by MeshCollider** [comment]
utACK

**Comment by sipa** [comment]
The code here looks like mostly output formatting changes, not actually dealing with the issue described by the title?

File: test/functional/test_runner.py [hunk 5]
```diff
@@ -40,17 +39,28 @@
...
 TEST_EXIT_PASSED = 0
 TEST_EXIT_SKIPPED = 77
 
+STATUS_PASSED = "Passed"
+STATUS_PASSED_WITH_WARNINGS = "Passed with warnings"
+STATUS_SKIPPED = "Skipped"
+STATUS_FAILED = "Failed"
+STATUSES = [STATUS_PASSED, STATUS_PASSED_WITH_WARNINGS, STATUS_SKIPPED, STATUS_FAILED]
+
+STATUS_MAX_LEN = max([len(st) for st in STATUSES])
```
Comment (ryanofsky) [inline test/functional/test_runner.py]: Square brackets unneeded and probably should be dropped for slightly better readability & efficiency.

**Review by ryanofsky — COMMENTED** [review]
utACK d64ac3f4aab13a1a4e8d320b52e7a2b934e011b2. Note that #10882 is currently depending on this so it needs 0.15 milestone.

> The code here looks like mostly output formatting changes, not actually dealing with the issue described by the title?

"Allow tests to pass" implies output formatting has to change at least somewhat to avoid giving impression  that tests failed when they actually succeeded, and these changes are a pretty straightforward way of updating the output.

**Comment by MarcoFalke** [comment]
utACK d64ac3f4aab13a1a4e8d320b52e7a2b934e011b2

**Comment by laanwj** [comment]
Reverted this in 60f9778 after discussion on IRC.
Sorry @MarcoFalke but there seems no agreement on doing this change.
It's unfortunate that #10882 requires this, but it seems like we'll have to look for an alternative solution there.

**Pull request merged**